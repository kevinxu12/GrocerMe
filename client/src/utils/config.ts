/**
 * @file Env constants for the frontend
 * @author Kevin Xu
 * TO DO - some of these variables should be dynamically calculated. We should figure out a way to clean this up.
 * For example,
 * - socket_url should depend on the environment i.e (dev, staging, prod). It should point to the backend node.js server, wherever that's hosted depending on the environment
 * - server_url should depend on the environment i.e (dev, staging, prod). ^ See above
 */
export const API_PREFIX = '/api';
export const environment =
  process.env.NODE_ENV ||
  process.env.IS_STAGING ||
  process.env.IS_PROD ||
  'development';

export const socket_url =
  environment === 'development'
    ? process.env.REACT_APP_PUBLIC_SERVER_DEV_URL || 'http://localhost:8080'
    : '/';

const server_url =
  environment === 'development'
    ? process.env.REACT_APP_PUBLIC_SERVER_DEV_URL || ''
    : ''; // by default dev url
console.log(server_url);
console.log(socket_url);
const server_api_url = server_url + API_PREFIX;
const calculated_configs = {
  server_url,
  server_api_url,
};

export default calculated_configs;

export const client_id = process.env.REACT_APP_GOOGLE_CLIENT_ID || '';
