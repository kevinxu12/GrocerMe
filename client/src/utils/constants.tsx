export enum Constants {
  DEFAULT_STRING = '',
  BLOB_PNG_URL = 'https://i.ibb.co/Hp5G1G4/blob-shape-1.png',
  HOME_URL = '/',
  USER_HOME = '/dashboard',
  TEST_URL = '/test',
}
