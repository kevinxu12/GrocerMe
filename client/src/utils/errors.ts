/**
 * @file Helpers for generating error messages in try / catches
 * @author Kevin Xu
 */
/* eslint-disable jsdoc/no-undefined-types */

type ErrorWithMessage = {
  message: string;
};

/**
 * Checks if is an error with a message
 *
 * @param {unknown} error Unknown error
 * @returns {boolean} true if error is with message, false otherwise
 */
function isErrorWithMessage(error: unknown): error is ErrorWithMessage {
  return (
    typeof error === 'object' &&
    error !== null &&
    'message' in error &&
    typeof (error as Record<string, unknown>).message === 'string'
  );
}
/**
 * Converts to a definitive error with a message
 *
 * @param {unknown} maybeError a potential error with a message
 * @returns {ErrorWithMessage} error with a message
 */
function toErrorWithMessage(maybeError: unknown): ErrorWithMessage {
  if (isErrorWithMessage(maybeError)) return maybeError;

  try {
    return new Error(JSON.stringify(maybeError));
  } catch {
    // fallback in case there's an error stringifying the maybeError
    // like with circular references for example.
    return new Error(String(maybeError));
  }
}
/**
 * Gets an Error Message from an unknown error
 *
 * @param {unknown} error the unknown error
 * @returns {string} error message
 */
export function getErrorMessage(error: unknown): string {
  return toErrorWithMessage(error).message;
}
