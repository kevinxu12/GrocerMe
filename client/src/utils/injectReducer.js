/**
 * @file Helpers to inject a new reducer into a store
 * @author Kevin Xu
 */
import React from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { ReactReduxContext } from 'react-redux';

import getInjectors from './reducerInjectors';

/**
 * Dynamically injects a reducer
 *
 * @param {object} props the props passed
 * @param {string} props.key the key to inject the reducer in
 * @param {object} props.reducer reducer to inject
 * @returns {Function} See below
 */
const wrappedComponent =
  ({ key, reducer }) =>
  WrappedComponent => {
    /**
     * A wrapper component that injects a given reducer into a key
     */
    class ReducerInjector extends React.Component {
      static WrappedComponent = WrappedComponent;

      static contextType = ReactReduxContext;

      static displayName = `withReducer(${
        WrappedComponent.displayName || WrappedComponent.name || 'Component'
      })`;

      /**
       * @param {object} props ReducerInjector Props
       * @param {object} context React Redux Context
       */
      constructor(props, context) {
        super(props, context);

        getInjectors(context.store).injectReducer(key, reducer);
      }

      /**
       * @returns {React.ElementType} Renders a Wrapped Component
       */
      render() {
        return <WrappedComponent {...this.props} />;
      }
    }

    return hoistNonReactStatics(ReducerInjector, WrappedComponent);
  };

export default wrappedComponent;

/**
 * Wrapper function to inject reducer without unnecessary renders
 *
 * @param {object} root0 general props
 * @param {string} root0.key the string of the reducer to inject
 * @param {object} root0.reducer the reducer object to inject
 */
const useInjectReducer = ({ key, reducer }) => {
  const context = React.useContext(ReactReduxContext);
  React.useEffect(() => {
    getInjectors(context.store).injectReducer(key, reducer);
  }, [context.store, key, reducer]);
};

export { useInjectReducer };
