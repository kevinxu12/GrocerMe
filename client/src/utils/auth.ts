/**
 * @file Generate authentication related helpers
 * @author Kevin Xu
 */

import { Role, RoleCode } from 'types/rest';
import { Constants } from './constants';

export const NO_ROLE_PATHS = [
  Constants.HOME_URL,
  Constants.TEST_SOCKET_URL,
  Constants.ABOUT,
  Constants.CONTACT_US,
];
const ALL_ROLE_PATHS = [Constants.SUPPLIER_HOME, Constants.USER_HOME];
const ADMIN_PATHS = [Constants.ADMIN_HOME];
const CONSUMER_PATHS = [];
const SUPPLIER_PATHS = [];
const role_path_object = {
  ADMIN: ADMIN_PATHS,
  CONSUMER: CONSUMER_PATHS,
  SUPPLIER: SUPPLIER_PATHS,
};

/**
 * Check if a given set of roles is authenticated for a path
 *
 * @param {string} path the path we want to check authentication for
 * @param {Role[]} roles the roles a user is permissioned to
 * @returns {boolean} whether the path is authenticated depending on the roles
 */
export const isPathAuthenticated = (path: string, roles: Role[]): boolean => {
  const role_codes = roles
    .filter((role: Role) => role.status)
    .map((role: Role) => role.code);
  // Admin, Supplier
  let all_paths = role_codes.map(code => role_path_object[code]);
  if (roles.length > 0) {
    all_paths = all_paths.concat(ALL_ROLE_PATHS);
  }
  const all_available_paths: string[] = NO_ROLE_PATHS.concat(
    [].concat(...all_paths),
  );
  return all_available_paths.includes(path);
};

/**
 * Check if a roleCode is contained within a list of roles
 *
 * @param {RoleCode} roleCode RoleCode that may or may not be included
 * @param {Role[]} roles roles to check within
 * @returns {boolean} true if included, false otherwise
 */
export const isRoleCodeIncluded = (
  roleCode: RoleCode,
  roles: Role[] | null,
): boolean => {
  if (!roles) {
    return false;
  }
  const matching_role = roles.filter((role: Role) => role.code === roleCode);
  return matching_role.length > 0;
};
