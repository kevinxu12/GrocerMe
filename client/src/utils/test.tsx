/**
 * @file Helper related to unit testing with jest
 */

import React, { ReactElement } from 'react';
import configureStore from 'redux-mock-store';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Store } from '@reduxjs/toolkit';
import { RoleCode } from 'types/rest';
import { Constants } from './constants';
import { AuthState } from 'types/RootState';
/* eslint-disable jsdoc/no-undefined-types */
/**
 * Generates a mock state for our auth reducer
 *
 * @param {RoleCode[]} roles the list of RoleCodes our mock logged in user has
 * @param {string} username the username of our mocked logged in user
 * @returns {AuthState} the mock auth state
 */
export const generateAuthState = (
  roles: RoleCode[],
  username: string = Constants.TEST_USERNAME,
): AuthState => {
  return {
    username: Constants.TEST_USERNAME,
    roles: roles,
  };
};

/**
 * A class that when instantiated, creates a mock provider and store for the entire testing class.
 * All subsequent components rendered with the helper methods access this store
 * Most wrappers will use <T> = <RootState>
 */
export class MockStoreWrapper<T> {
  mockStore: any; // not sure what's the typing here
  state: T; // the store state

  /**
   * Initialize the Mock Store Wrapper
   *
   * @param {T} state The initial state for the store
   */
  constructor(state: T) {
    this.mockStore = configureStore();
    this.state = state;
  }

  /**
   * Create a mock store with an explicit state, default to the state instance variable otherwise
   *
   * @param {T} state an explicit state to keep track of
   * @returns {Store} a mock store that holds the desired state
   */
  public generateStore(state: T = this.state): Store<T> {
    return this.mockStore(state);
  }

  /**
   * Wraps a component with a Provider and Router
   *
   * @param {React.ReactElement} component Component to be wrapped
   * @param {T} state Optional state to create store with
   * @returns {RenderResult} testing library render
   */
  public renderComponentWithRouterProvider(
    component: React.ReactElement,
    state: T = this.state,
  ): RenderResult {
    const store = this.generateStore(state);
    const jsx = (
      <Provider store={store}>
        <Router>{component}</Router>
      </Provider>
    );
    return render(jsx);
  }

  /**
   * Wraps a component with a Provider and Router
   *
   * @param {React.ReactElement} component Component to be wrapped
   * @param {T} state Optional state to create store with
   * @returns {ReactElement} testing library render
   */
  public genComponentWithRouterProvider(
    component: React.ReactElement,
    state: T = this.state,
  ): React.ReactElement {
    const store = this.generateStore(state);
    const jsx = (
      <Provider store={store}>
        <Router>{component}</Router>
      </Provider>
    );
    return jsx;
  }

  /**
   * Wraps a component with a Provider
   *
   * @param {React.ReactElement} component Component to be wrapped
   * @param {T} state Optional state to create store with
   * @returns {RenderResult} testing library render
   */
  public renderComponentWithProvider(
    component: React.ReactElement,
    state: T = this.state,
  ): RenderResult {
    const store = this.generateStore(state);
    return render(<Provider store={store}>{component}</Provider>);
  }
}

/**
 * Test Component for class
 *
 * @returns {React.ReactElement} Test Component
 */
export const TestComponent = (): React.ReactElement => {
  return <div> Test </div>;
};
