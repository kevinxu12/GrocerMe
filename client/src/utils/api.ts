/**
 * @file An Api client wrapper around axios. Use this instead of calling axios directly
 * @author Kevin Xu
 */
import axios, { AxiosResponse } from 'axios';
import { getErrorMessage } from './errors';
import generateServerUrl from './url';
/**
 *
 */
class Api {
  /**
   * Wrapper around axios.get
   *
   * @param {string} url the raw url string
   * @param {Function} cb Optional callback to run on error
   * @param {string} success_message Success message
   * @returns {AxiosResponse} axiosResponse
   */
  public static async get(
    url: string,
    cb?: Function,
    success_message?: string,
  ) {
    try {
      const res = await axios.get(generateServerUrl(url), {
        withCredentials: true,
      });
      if (cb && success_message) {
        await cb(success_message);
      }
      return res;
    } catch (error) {
      if (cb) {
        cb(getErrorMessage(error));
      } else {
        throw error;
      }
    }
  }

  /**
   * Wrapper around axios.post
   *
   * TO DO, we should probably type the request body, instead of using ANY in the future.
   *
   * @param {string} url raw url string
   * @param {any} body Post object
   * @param {Function} cb Optional Callback
   * @param {string} success_message success message to call callback with
   * @returns {AxiosResponse} axiosResponse
   */
  public static async post(
    url: string,
    body: any = {},
    cb?: Function,
    success_message?: string,
  ) {
    try {
      const res = await axios.post(generateServerUrl(url), body, {
        withCredentials: true,
      });
      if (cb && success_message) {
        await cb(success_message);
      }
      return res;
    } catch (error) {
      if (cb) {
        cb(getErrorMessage(error));
      } else {
        throw error;
      }
    }
  }
}
export default Api;
