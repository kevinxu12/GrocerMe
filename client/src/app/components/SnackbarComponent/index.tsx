/**
 * @file Component Wrapper that makes it easy to display any snackbar API alerts
 * Adds a prop called setMessage to the child component that specifies the snackbar message
 * @author Kevin Xu
 */
import React, { useCallback, useState } from 'react';
import { isErrorMessage } from 'utils/errors';
import { CustomSnackBar } from '../Snackbar';

/**
 * See file description
 *
 * @param {Object} param0 General Props
 * @param {React.ReactElement} param0.component Component to wrap around
 * @returns {React.ReactElement} A wrapped component
 */
const SnackbarComponent = ({
  component: Component,
  ...rest
}): React.ReactElement => {
  const [message, setMessage] = useState<string>('');
  const [isReset, setIsReset] = useState<boolean>(true);
  /**
   * resets the component to display a new snackbar
   * Unsure if useCallback is needed here.
   */
  const reset = useCallback(() => {
    setIsReset(true);
    setMessage('');
  }, []);

  return (
    <div>
      <Component {...rest} setMessage={setMessage} />
      {message && isErrorMessage(message) && isReset && (
        <CustomSnackBar text={message} reset={reset} severity={'error'} />
      )}
      {message && !isErrorMessage(message) && isReset && (
        <CustomSnackBar text={message} reset={reset} severity={'success'} />
      )}
    </div>
  );
};

export default SnackbarComponent;
