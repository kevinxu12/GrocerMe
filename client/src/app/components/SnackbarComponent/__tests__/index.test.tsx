import * as React from 'react';
import SnackbarComponent  from '..';
import renderer from 'react-test-renderer';
import { TestComponent } from 'utils/test';

const renderPage = () =>
  renderer.create(
      <SnackbarComponent component = {TestComponent}/>
  );

describe('<Snackbar Component  />', () => {
  
    it('Snackbar wrapper default', () => {
      const sw = renderPage();
      expect(sw.toJSON()).toMatchSnapshot();
    });
  });