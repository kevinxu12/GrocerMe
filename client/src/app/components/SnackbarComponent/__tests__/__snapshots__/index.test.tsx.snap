// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotFoundPage  /> Snackbar wrapper default 1`] = `
<div>
  <div>
     Test 
  </div>
  
  
</div>
`;
